src/singledocparser.cpp

```diff
@@ -46,6 +46,9 @@ void SingleDocParser::HandleDocument(EventHandler& eventHandler) {
 }
 
 void SingleDocParser::HandleNode(EventHandler& eventHandler) {
+  if (depth > depth_limit) {
+    throw ParserException(m_scanner.mark(), ErrorMsg::BAD_FILE);
```

Comment:
> Note that throwing will eliminate the possibility of re-using this object because it does not reset depth.

Is that a serious issue? At this point - the file should be considered garbage anyways and we just need to abort gracefully.

> I suggest using a depth guard type of class and instantiate that at the entrypoint of the function. A quick example: [...]

That's pretty neat - didn't think of using destructors to take care of this... But then that's a huge change - at that point, I would rather switch to a non-recursive design, as you said, but that's also a large change.

I'm looking at making a rather non-invasive change here, that can be backported to earlier versions as well. I'm also worried about stack size problems here - 2000 may just be too high as well - having an arbitrary number there is just looking for trouble. But I'm not sure how else to fix this, short of rewriting the whole parsing logic to be non-recursive.